README.md
Advanced Wesnoth Wars
=======================

About
-----------

It's a mod  for "Battle for Wesnoth" open-source videogame.

* Author :  [Ruvaak](http://thomas.bondois.info)
* Version : see [lua/aww_status.lua:9](lua/aww_status.lua)
* Changelog : see [CHANGELOG.md](CHANGELOG.md)
* License : [GNU-GPL](LICENSE.md)
* Code Repository [on GitHub](https://github.com/tbondois/Advance_Wesnoth_Wars)
* Dedicated [forum thread](https://forums.wesnoth.org/viewtopic.php?f=15&t=49454)


DISCLAIMER FOR MODDERS
----------------------

You can reuse the code of this mod in your own creations, without asking my authorization, only if you share it with the same License, freely.
But I will really appreciate to know my code is used on your project, and that you credit my work somewhere in your code.


Description
-----------

Last Wesnoth version tested : 1.14.7

A collection of the most-requested options to make the mechanics more realistic and tactic, for people curious to experiment a new gameplay.
It changes the combat approach, making it easier for good tacticians, harder for the others.
All theses features works for mainline Campaigns & Scenarios, Single or Multiplayer, for all sides (including AI)
It also includes some options to give interest to level-up units after max advancement.

The features which keep the Campaigns balanced are enabled by default. The others features are more aimed for PvP multiplayer, or being played on campaign in a lesser difficulty.

Features list :

- No Random Combats : attacks never miss, terrain bonus is instead used as damage reducer.

- Increased Damage : To compensate the high terrains defenses. Only applies for 'No Random Combats'.

- Squad Mode (HP-related strikes) : Units lacking lot of HP do proportionally less attack strikes (or less damage for single-strike attacks). Swarm is a worse ratio (-1 strike if you miss just 1 HP), but native, so considered by AI. In both cases, minimum strikes will be 1, and will not apply for berserk.

- Level-Up : Relative Healing : When an unit advance a level, instead of a full healing and being cured, it keep the missing HP, but statuses are cured. The special case when post-advancement Max-HP is smaller than pre-advanced is also handled.

- Level-Up : Random Bonuses After Max Level Advancement : Units already AMLA (purple XP bar) gain an extra random ability or increased stat each time they level-up again.

- Level-Up : Promoted Leaders After Max Level Advancement : Post-level 3 standard units reaching their first AMLA level-up will be able to recruit the same units as the original leader, recall, and be prefixed 'Chief' with a bronze crown icon (keeping the Loyal icon in case). Excluded for special Heroes.

- *Updated* Epic Heroes : The Leaders and Heroes (gold & silver crowns) gain an increase of 20% of their attributes and a small regeneration ability, making them harder to kill. Excluded for Promoted Leaders.

- Learning from battlefield (Passive XP) : Extra XP for all units each turn (except for the ones not recalled).

- Learning from healing (Max XP/turn for Healing) : Each turn, Healers will earn 1 XP for each adjacent wounded (but not-poisoned) ally, but you can limit the max/turn.

- *Updated* Berserk tweak - Fury : Replace 'berserk' by a new 'Fury' weapon special, way more interesting. A warcry related to the fury level with also be displayed.

- *Updated* Ambushed tweak - Surprise Attacks : An ambush will trigger a quick combat in which the ambushed unit can't counter-attack. Awesome combined with NINJA WARS!

- *NEW* Ninja Wars (Stealthy units) : Most of units will melee back-stabs, poison arrows, distract from ZoC, be faster and invisible/ambushers in villages, forests, deep water, or everywhere at night. For funny PvP with fog-of-war focused on ambushes / hide&seek.

- *NEW* Level-Up After Max Level Advancement : Increase Level Number (The unit level will continue growing with the AMLA level-ups)

- Also includes a PvPvE scenario : '4p - Ruvaak Mirage Atoll' optimized to be combined with all features.

Feel free to contribute !
If you experiment a bug, please report it to me on the forum, then try to disable the feature (see next section).


How To Enable/Disable Mod Options during a Scenario/Campaign
-----------------------------------------------

The mod works when features options are changed during a game.
When you disable an option, *MOST* of the changes made on units by the features are removed. But it's experimental.

1) During a game, type :
```
:debug
``` 
(except if you launched the game with --debug argument)
2) Then open Lua Console(default : `, but you can see it in Menu > Settings > Shortcuts > Display lua console)
3) Copy/paste/execute one of the following lines related to mod features. 
    - Change the number in parentheses : 0 (zero) means disabled, 1 enabled.
    - When specified, you can replace it by another number. (like in feature 02, 04, 05, 08...)
    - Most of the changes will be operational immediately, or next turn, or next scenario. But units already modified will keep theirs changes.
The commands to edit each feature and remove/re-enable changes on units :
```lua
aww_status.update_feature_01(0) -- NoRandomCombats
aww_status.update_feature_02(0) -- Squad Mode 1 = custom, 2 = swarm
aww_status.update_feature_03(0) -- L-Up AMLA Promoted Leader
aww_status.update_feature_04(0) -- passive xp, 0 to 6
aww_status.update_feature_05(0) -- healing xp 0 to 6
aww_status.update_feature_06(0) -- L-Up  relative healing
aww_status.update_feature_07(0) -- verbose debug messages
aww_status.update_feature_08(0) -- NoRandomCombats Damages Adjustment, to -20 to 40
aww_status.update_feature_09(0) -- AMLA Random Bonuses
aww_status.update_feature_10(0) -- Epic Leaders
aww_status.update_feature_11(0) -- Ninja Wars
aww_status.update_feature_12(0) -- Berserk tweak
aww_status.update_feature_13(0) -- Ambush tweak
aww_status.update_feature_14(0) -- L-Up Notif
aww_status.update_feature_15(0) -- L-Up AMLA number
```
If you want to display all currently enabled features (and the associated number), copy/paste/execute this line the lua console :
```lua
aww_status.message_info_all()
```


How To Migrate a Savegame from a previous version of the mod
-------------------------------

It will do it automatically on next scenario. If you want to try to migrate the current scenario, copy/paste in lua console the following line :
```lua
wesnoth.fire_event('aww_event_reload_lua')
```
If you see some bugs on a feature,  (see previous section) disable it on re-enable it (replace XX by feature number):
```lua
aww_status.update_feature_XX(0)
aww_status.update_feature_XX(1)
```
See previous section to know how to open console or when the changes will applies.



Developer Mode
--------------

This is useful only for development/test purpose on the addon lua scripts. 

- Do not do that play a big campaign, as it can create issue with your savegame when you update the addon.
- Effect : the LUA scripts will be loaded externally (means you can edit them, and the effect will apply when you reload a savegame, no need to re-create a game or a scenario like for WML scripts).
- To enable it, just place a empty file named :
```
aww.dev
```
in the addon base directory (where the _main.cfg is).

- To disable it, just remove the file, and restart the game.


Thanks
-----------
Some of theses features are inspired from existing code, saving me lot of times.
A big thanks particularly to Eagle 11, Ravana, Bob_The_Mighty, Dugi, for their creation.
And a huge thanks for all developers and Wesnoth WML wiki contributors !


CHANGELOG.md
# Changelog

For more clarity about compatibility, first 2 version numbers will follow base game version.

So 1.14.X.Y means game was checked compatible with Wesnoth 1.14. X is incremented for new features ; Y is increment for fixes/optimisations.

## [1.14.15.1] - 2019-10-19

### Added
- Feature #01 : damage table on attack description
- Feature #01 : special hit 100% not cumulative

### Changed
- hit chance on special attack (magic etc) will be the minimum and not the absolute

## [1.14.14.5] - 2019-05-04

### Fixed
- Berserk tweak : reset ratio/kills/bloodthirsty when unit is recalled/next scenario

## [1.14.14.4] - 2019-04-20

### Changed
- Some float text colors
- Colors or trait name epic (gold), squad swarm special (cyan) on units properties.
- All traits (epic, ninja) and abilities/weapons-special (fury, bloodthirsty, distract, camouflage, custom regen...) defined in the mod are in italic on unit properties, to make a distinction from what was added by the mod.

### Fixed
- Optimised disabling of Feature #12 Berserk Tweak

### Deleted
- old lua/aww_combats.lua

## [1.14.14.3] - 2019-04-16

### Fixed
- Bug reported by Hammerfritz where hit chance was put in 100% in the case "No Random Combats" was disabled, but "Squad Mode" was.
- Bug on "#08 : damage adjustement" not always taken in account
- Bug aww_status.migrate() was always executed.

## [1.14.14.2] - 2019-04-14

### Added
- Feature #12 Berserk Tweak : dwarf laugh when a fury-wearer unit kill someone.

### Changed
- Feature #11, ninja, #10 Epic : found a way to enable / remove trait
- re-enabled "stealthy" trait (renamed "ninja") for Feature #11
- Feature #06 : Level-Up : Relative Healing disabled by default, as it can break balance of campaigns.

### Fixed
- Minor error on migrating version of a savegame
- Bad display information for Passive XP on chat message.
- Bloodthirsty was added after first kill instead of 5th on feature #12 Berserk tweak

## [1.14.14.1] - 2019-04-13

### Changed
- re-balanced Feature #12 (Berserk tweak) : no more initial offensive drains, base attacks damage don't increase on kills, but after 5 kills the fury attack gains also Bloodthirsty (offensive drains)
- Feature #11, ninja : movement increase is used in [object] like the others changes,  don't include a 'stealthy' trait anymore.
- improved aww_status.update_feature_07()  and update_feature_08()
- Re-balance feature #9 AMLA bonus : in case it get the bloodthirsty special attack, apply only for current scenario

### Fixed
- Minor error on migrating version of a savegame
- Bad display information for Passive XP on chat message.

## [1.14.14.0] - 2019-04-13 (BfW 1.14.7 released)

### Added
- Feature #15 : Level-Up After Max Level Advancement : Increase Level Number (default ON)
- Explanations in README.md and code adaption (aww_status.lua, aww_duel.lua) to enabled/disable feature during a game/campaign
- aww_status.run.lua (to separate aww_status declaration from execution)
- aww_status.lua : some functions to manage disable/enable features
- events _enable _restart, _reset on most features impacting units, to revert / re-enable changes (experimental)

### Changed
- Feature #13 : Ambush Tweak no longer in beta, added hides cooldown of 1 turn after surprise/normal attack
- Options numbers are now the same as feature numbers.
- Squad Mode Custom : strikes special effect is now a cumulative multiplier, instead of a not-cumulative
- Squad Mode : Swarm : attacks having 1 base strike will do reduced HP-related damages, like in Custom Mode.
- NoRandomCombats & Squad Mode : added + 0.01 to multiplier to prevent a round down and get a round(.5)
- Feature "Gifted heroes" renamed "Epic heroes"
- balancing Epic heroes : don't include anymore "leadership" ability (already give custom regen 4/slowed + distract)
- Some adjustments on gifted trait stats, and renamed it "epic"
- Max attack damage points earned with "Fury" changed from 15 to 10
- Somes changes in scenario + map 4p_Ruvaak_Mirage_Atoll (now v1.4)
- gettext : textdomains wesnoth-aww and wesnoth-aww-rma merged in one "aww" (and scenario translation reworks)
- macro files *.ma.cfg renamed *.mac.cfg
- uses of [objects] in some features, for effects/abilities
- Features #09 AMLA Random Bonuses : less probabiities to have moves +1

### Fixed
- LUA error sometimes in the end of scenario when NoRandom-Combats/Squad Mode Custom is enabled and no side is defined : (Saw Voyage of a drake > skip tutorial)
- Healing XP : apply to ability-type "heals" instead of only ability id "healing". Optimized not-poisoned filter.

### Deleted
- useless about.cfg

## [1.14.13.0] - 2019-04-04

### Added
- Ability 'distract' for "Gifted Leader". It's a Skirmisher leadership from data/campaigns/The_Rise_Of_Wesnoth/utils/trow-abilities.cfg
- Ability 'leadership' for "Gifted Leader"

### Changed
- Completely changed the "berserk tweak"
- Update french translations
- Ability 'distract' instead of 'skirmisher' for "NINJA WARS!" & "ALMA Random Bonus"
- ability name for camouflage changed from "ambush" to "aww_ability_camouflage"

### Fixed
- Issue reported by Hammerfriz: Squad Mode Custom was not working sometimes (special set to cumulative=false)


## [1.14.12.6] - 2019-03-31

### Changed
- "10. Berserk tweak : Fury" becomes : "10. Berserk tweak : Fury & Drains & Drains"
- "7. Epic heroes" : + leadership ability

## [1.14.12.5] - 2019-03-31

### Added
- Estimations of damages on each weapons (No Random Combats or Squad Mode except if Squad Mode : Swarm chosen)
    - on events : select,sighted,advance,recruit, attack end (for both units)
    - Note : "select" do not apply on enemy unit.
- translation strings in LUA script
- 100% french translated ! Cocorico !

### Changed
- some variables names in aww_duel.lua : notion of constants uppercases
- some french trads

### Fixed
- clear aww_bonus_rand variable after each use (optimisation) (09_amla_bonus)
- lua errors on hit_chance calculation with marksmanshit units (No Random Combats)


lua/aww_status.lua
--<<
-- Author: Ruvaak
-- Original Creation

local _ = wesnoth.textdomain 'aww'

aww_status = {
	title = _"Advanced Wesnoth Wars",
	version = "1.14.14.5",
	FEATURE_IDS = {
		[1]  = "aww_01_enable_randomless_combats",
		[2]  = "aww_02_squad_mode",
		[3]  = "aww_03_enable_promoted_leader",
		[4]  = "aww_04_passive_xp",
		[5]  = "aww_05_healing_xp",
		[6]  = "aww_06_enable_levelup_relative_healing",
		[7]  = "aww_07_enable_verbose",
		[8]  = "aww_08_randomless_damage_adjustment",
		[9]  = "aww_09_enable_amla_bonus",
		[10] = "aww_10_enable_gifted_leaders",
		[11] = "aww_11_enable_ninja",
		[12] = "aww_12_enable_berserk_fix",
		[13] = "aww_13_enable_ambushed_fix",
		[14] = "aww_14_enable_levelup_notif",
		[15] = "aww_15_enable_levelup_amla_inc",
	},
}


function aww_status.run()
	aww_status.versioning()
	aww_status.init()
	aww_status.message_info()
end


function aww_status.init()
	aww_status.feature_01 = aww_status.get_feature_value(1, false)
	aww_status.feature_02 = aww_status.get_feature_value(2, 0)
	aww_status.feature_03 = aww_status.get_feature_value(3, false)
	aww_status.feature_04 = aww_status.get_feature_value(4, 0)
	aww_status.feature_05 = aww_status.get_feature_value(5, 0)
	aww_status.feature_06 = aww_status.get_feature_value(6, false)
	aww_status.feature_07 = aww_status.get_feature_value(7, false)
	aww_status.feature_08 = aww_status.get_feature_value(8, 0)
	aww_status.feature_09 = aww_status.get_feature_value(9, false)
	aww_status.feature_10 = aww_status.get_feature_value(10, false)
	aww_status.feature_11 = aww_status.get_feature_value(11, false)
	aww_status.feature_12 = aww_status.get_feature_value(12, false)
	aww_status.feature_13 = aww_status.get_feature_value(13, false)
	aww_status.feature_14 = aww_status.get_feature_value(14, false)
	aww_status.feature_15 = aww_status.get_feature_value(15, false)
end


function aww_status.versioning()
	local prev_version = wesnoth.get_variable("aww_version")
	if prev_version ~= nil and prev_version ~= aww_status.version then
		aww_status.migrate(prev_version)
	end
	wesnoth.set_variable("aww_version", aww_status.version)
end


-- migrate old versions name to keep options through versions of AWW on same campaign :
function aww_status.migrate(from_version)
	if aww_status.get_feature_value(1, nil) == nil and wesnoth.get_variable("aww_enable_randomless_combats") ~= nil then
		aww_status.update_feature_01(wesnoth.get_variable("aww_enable_randomless_combats"))
	end
	if aww_status.get_feature_value(2, nil) == nil and wesnoth.get_variable("aww_enable_health_based_combats") ~= nil then
		local was_enabled_02 = wesnoth.get_variable("aww_enable_health_based_combats")
		if was_enabled_02 == true then
			aww_status.update_feature_02(1)
		else
			aww_status.update_feature_02(0)
		end
	end
	if aww_status.get_feature_value(3, nil) == nil and wesnoth.get_variable("aww_enable_leader_promotions")~= nil then
		aww_status.update_feature_03(wesnoth.get_variable("aww_enable_leader_promotions"))
	end
	if aww_status.get_feature_value(4, nil) == nil and wesnoth.get_variable("aww_passive_xp") ~= nil then
		aww_status.update_feature_04(wesnoth.get_variable("aww_passive_xp")*1)
	end
	if aww_status.get_feature_value(5, nil) == nil and wesnoth.get_variable("aww_healing_xp") ~= nil  then
		aww_status.update_feature_05(wesnoth.get_variable("aww_healing_xp")*1)
	end
	if aww_status.get_feature_value(6, nil) == nil and wesnoth.get_variable("aww_enable_advance_healing_reduced") ~= nil then
		aww_status.update_feature_06(wesnoth.get_variable("aww_enable_advance_healing_reduced"))
	end
	if aww_status.get_feature_value(7, nil) == nil and wesnoth.get_variable("aww_enable_debug") ~= nil  then
		aww_status.update_feature_07(wesnoth.get_variable("aww_enable_debug"))
	end
	if aww_status.get_feature_value(8, nil) == nil and wesnoth.get_variable("aww_randomless_terrain_adjustment") ~= nil then
		aww_status.update_feature_08(wesnoth.get_variable("aww_randomless_terrain_adjustment")*1)
	end
	if aww_status.get_feature_value(9, nil) == nil and wesnoth.get_variable("aww_enable_amla_extra_bonus") ~= nil  then
		aww_status.update_feature_09(wesnoth.get_variable("aww_enable_amla_extra_bonus"))
	end
	if aww_status.get_feature_value(10, nil)== nil and wesnoth.get_variable("aww_enable_gifted_leaders") ~= nil  then
		aww_status.update_feature_10(wesnoth.get_variable("aww_enable_gifted_leaders"))
	end
	if aww_status.get_feature_value(11, nil) == nil and wesnoth.get_variable("aww_enable_stealth_mode") ~= nil  then
		aww_status.update_feature_11(wesnoth.get_variable("aww_enable_stealth_mode"))
	end

	wesnoth.message("AWW notice", string.format("migrating from version %s to %s", wesnoth.get_variable("aww_version"), aww_status.version))
	wesnoth.set_variable("aww_migrated_version", from_version)
end


function aww_status.get_info(filter_id)
	-- filter_id is an optionnal parameter. if not present/nil, will display all enabled features in the same order as in options menu
	local msg = ""
	if (filter_id == nil or filter_id == 1) and aww_status.feature_01 then
		msg = "01. ".._"No Random Combats" .. " | "
	end
	if aww_status.feature_01 and (filter_id == nil or filter_id == 1) and aww_status.feature_08 ~= 0 then
		msg = msg .. "08. ".._"Increased Damage"
				.. string.format(" [%s%%]",aww_status.feature_08)
				.. " | "
	end
	if (filter_id == nil or filter_id == 2) and aww_status.feature_02 and aww_status.feature_02 ~= 0 then
		local label = ""
		if aww_status.feature_02 == 1 then
			label = _"Custom"
		elseif aww_status.feature_02 == 2 then
			label = _"Swarm"
		end
		msg = msg .. "02. ".._"Squad Mode"
				.. string.format(" [%s:", aww_status.feature_02)
				..  label
				.. "] | "
	end
	if (filter_id == nil or filter_id == 6) and aww_status.feature_06 then
		msg = msg .. "06. ".. _"L-Up ".. _"Relative Heal" .. " | "
	end
	if (filter_id == nil or filter_id == 14) and aww_status.feature_14 then
		msg = msg .. "14. ".. _"L-Up ".._"Notif" .. " | "
	end
	if (filter_id == nil or filter_id == 15) and aww_status.feature_15 then
		msg = msg .. "15. L-Up ".._"AMLA".." ".. _"Increase Level Number" .. " | "
	end
	if (filter_id == nil or filter_id == 9) and aww_status.feature_09 then
		msg = msg .. "09. ".. _"L-Up ".._"AMLA".." ".._"Bonuses" .. " | "
	end
	if (filter_id == nil or filter_id == 3) and aww_status.feature_03 then
		msg = msg .. "03. ".. _"L-Up ".._"AMLA".." ".._"Promoted Leaders" .. " | "
	end
	if (filter_id == nil or filter_id == 10) and aww_status.feature_10 then
		msg = msg .. "10. ".. _"Epic Heroes" .. " | "
	end
	if (filter_id == nil or filter_id == 4) and aww_status.feature_04 and aww_status.feature_04 ~= 0 then
		msg = msg .. "04. ".. _"Passive XP"
		 .. string.format(" [%s", aww_status.feature_04 )
		.. _"/turn]"
		 .. " | "
	end
	if (filter_id == nil or filter_id == 5) and aww_status.feature_05 and aww_status.feature_05 ~= 0
	then msg = msg .. "05. ".._"Healing XP"
			.. _" [Max "
			.. string.format("%s", aww_status.feature_05)
			.. _"/turn]"
			.. " | "
	end
	if (filter_id == nil or filter_id == 12) and aww_status.feature_12 then
		msg = msg .. "12. ".._"Berserk tweak" .. " | "
	end
	if (filter_id == nil or filter_id == 13) and aww_status.feature_13 then
		msg = msg .. "13. ".._"Ambush tweak" .. " | "
	end
	if (filter_id == nil or filter_id == 11) and aww_status.feature_11 then
		msg = msg .. "11. ".._"NINJA WARS!" .. " | "
	end
	if (filter_id == nil or filter_id == 7) and aww_status.feature_07 then
		msg = msg .. "07. ".._"Verbose"
	end

	if msg ~= "" then
		msg = _"Enabled :" .. " " .. msg
	elseif filter_id then
		msg = aww_status.get_feature_info(filter_id)
	else
		msg = _"All mod options are disabled."
	end
	return msg
end


function aww_status.get_feature_info(id)
	local msg = string.format(" #%s. %s = [%s]", id, aww_status.feature_code(id), aww_status.get_feature_value(id, '?'))
    msg = string.gsub(msg, "true", "ON")
    msg = string.gsub(msg, "false", "off")
    msg = string.gsub(msg, "aww_", "")
    msg = string.gsub(msg, "_", " ")
	return _"Feature" .. msg
end


function aww_status.message_info(id)
	wesnoth.message(aww_status.title .. " v".. aww_status.version, aww_status.get_info(id))
end


function aww_status.message_info_all()
	local msg = ""
	for id = 1, #aww_status.FEATURE_IDS do
		msg = msg .. aww_status.get_feature_info(id) .. " | "
	end
	wesnoth.message(aww_status.title .. " v".. aww_status.version, msg)
end


function aww_status.feature_code(id)
	if aww_status.FEATURE_IDS[id] ~= nil then
		return aww_status.FEATURE_IDS[id]
	end
	return "undefined"
end

function aww_status.get_feature_value(id, nil_value)
	local value = wesnoth.get_variable(aww_status.feature_code(id))
	if value == nil then
		 return nil_value
	end
	return value
end

function aww_status.set_feature_value(id, value)
	wesnoth.set_variable(aww_status.feature_code(id), value)
end

function aww_status.update_feature_01(value)
	local id = 1
	value = aww_status.to_bool(value)
	aww_status.feature_01 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_01_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
		wesnoth.fire_event("aww_event_01_02_reset")
		if value then
			wesnoth.fire_event('aww_event_reload_duel')
		end
	end
	return value
end


function aww_status.update_feature_02(value)
	local id = 2
	value = value *1
	aww_status.feature_02 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_02_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
		wesnoth.fire_event("aww_event_01_02_reset")
		if old_value == 0 then
			wesnoth.fire_event('aww_event_reload_duel')
		end
		if old_value == 2 then
			wesnoth.fire_event("aww_event_02_swarm_disable")
		end
		if value == 2 then
			wesnoth.fire_event("aww_event_02_swarm_enable")
		end
	end
	return value
end


function aww_status.update_feature_03(value)
	local id = 3
	value = aww_status.to_bool(value)
	aww_status.feature_03 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_03_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
		if value then
			wesnoth.fire_event("aww_event_03_enable")
		else
			wesnoth.fire_event("aww_event_03_disable")
		end
		if value == false then
		end
	end
	return value
end

function aww_status.update_feature_04(value)
	local id = 5
	value = value *1
	aww_status.feature_04 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_04_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
	end
	return value
end

function aww_status.update_feature_05(value)
	local id = 5
	value = value *1
	aww_status.feature_05 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value  then
		aww_status.feature_05_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
	end
	return value
end

function aww_status.update_feature_06(value)
	local id = 6
	value = aww_status.to_bool(value)
	aww_status.feature_06 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_06_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
	end
	return value
end

function aww_status.update_feature_07(value)
	local id = 7
	value = aww_status.to_bool(value)
	aww_status.feature_07 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_07_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
		if (aww_status.feature_01 or aww_status.feature_02 > 0) then
			wesnoth.fire_event('aww_event_reload_duel')
		end
	end
	return value
end

function aww_status.update_feature_08(value)
	local id = 8
	value = value *1
	aww_status.feature_08 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_08_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
		if (aww_status.feature_01) then
			wesnoth.fire_event('aww_event_reload_duel')
			wesnoth.fire_event("aww_event_01_02_reset")
		end
	end
	return value
end

function aww_status.update_feature_09(value)
	local id = 9
	value = aww_status.to_bool(value)
	aww_status.feature_09 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_09_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
		if value == false then
			wesnoth.fire_event("aww_event_09_disable")
		end
	end
	return value
end

function aww_status.update_feature_10(value)
	local id = 10
	value = aww_status.to_bool(value)
	aww_status.feature_10 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_10_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
		if value then
			wesnoth.fire_event("aww_event_10_enable")
		else
			wesnoth.fire_event("aww_event_10_disable")
		end
	end
	return value
end

function aww_status.update_feature_11(value)
	local id = 11
	value = aww_status.to_bool(value)
	aww_status.feature_11 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_11_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
		if value then
			wesnoth.fire_event("aww_event_11_enable")
		else
			wesnoth.fire_event("aww_event_11_disable")
		end
	end
	return value
end


function aww_status.update_feature_12(value)
	local id = 12
	value = aww_status.to_bool(value)
	aww_status.feature_12 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_12_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
		if value then
			wesnoth.fire_event("aww_event_12_enable")
		else
			wesnoth.fire_event("aww_event_12_disable")
		end
	end
	return value
end


function aww_status.update_feature_13(value)
	local id = 13
	value = aww_status.to_bool(value)
	value = aww_status.to_bool(value)
	aww_status.feature_13 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_13_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
		wesnoth.fire_event("aww_event_13_reset")
	end
	return value
end

function aww_status.update_feature_14(value)
	local id = 14
	value = aww_status.to_bool(value)
	aww_status.feature_14 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_14_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
	end
	return value
end

function aww_status.update_feature_15(value)
	local id = 15
	value = aww_status.to_bool(value)
	aww_status.feature_15 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_13_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
	end
	return value
end


function aww_status.to_bool(v)
	if (type(v) =="boolean") then
		return v
	end
    return v and (
		(type(v) =="number" and (v > 0))
		or (type(v)=="string" and v ~= "false" and v ~= "no" and v ~= "0" and v ~= "")
	)
end


-- DEPRECATED alias for run(), for retrocompatibility in dev mode :
function aww_status.welcome()
	aww_status.run()
end


-- CUSTOM WML TAGS :

-- used in mods/10_gifted_leader_hero.cfg, mods/11_ninja.cfg
function wesnoth.wml_actions.aww_remove_trait(cfg)
	local trait_id = cfg.trait_id
	for _,unit in ipairs(wesnoth.get_units(cfg)) do
		wesnoth.remove_modifications(unit, {id = trait_id}, "trait")
	end
end

-- aww_status.run() -- load aww_status.up.lua instead
-->>
